### PR TITLE
feat(rush-lib): two stage install & --ignore-scripts

### PR DESCRIPTION
**Pull request by chengcyber**
## Summary

This PR implements "two stage install" proposal.
close https://github.com/microsoft/rushstack/issues/3530

## Details

Workflow now:

1. Divide "rush install/update" into two stage install when using pnpm & "useWorkspaces" on
2. During two stage install, '--ignore-scripts' are pushed as a cli parameter to "pnpm install"
3. After "pnpm install" successfully, if "--ignore-scripts" is not specified explicitly by user, run "pnpm rebuild --pending"

Goal: faster retries for install lifecycle scripts:
If step3 fails, such as encountering an error in a "postinstall" scripts, rerun "rush install" will skip step2, which gets a faster retry on failed scripts.

## How it was tested

Use rushstack repo for example

1. `rush --debug install -t rush`

2022-07-14 at 3 29 PM

two stage install works :)

2. `rush --debug install -t rush-lib`

Because `-t rush-lib` is a subset of `-t rush`, stage1(pnpm install w/ --ignore-scripts) was skipped, and stage2(pnpm rebuild --pending) was run.

2022-07-14 at 3 32 PM

### CODE REVIEW

File: libraries/rush-lib/src/logic/base/BaseInstallManager.ts [hunk 1]
```diff
@@ -154,11 +177,26 @@ export abstract class BaseInstallManager {
...
+  protected get twoStageInstall(): boolean {
+    return this._twoStageInstall;
+  }
+
   public async doInstallAsync(): Promise<void> {
     const isFilteredInstall: boolean = this.options.pnpmFilterArguments.length > 0;
     const useWorkspaces: boolean =
       this.rushConfiguration.pnpmOptions && this.rushConfiguration.pnpmOptions.useWorkspaces;
 
+    this._twoStageInstall = useWorkspaces && this.rushConfiguration.packageManager === 'pnpm';
```
Comment (octogonz) [inline libraries/rush-lib/src/logic/base/BaseInstallManager.ts:193]: For the non-draft PR, we should provide a setting to restore the old behavior, in case people encounter any problems with this new mode.  

An optimistic approach might be `legacyPostinstallBehavior=true` (opt-out) in **experiments.json**.  

A less optimistic approach might be `deferredPostinstall=true` (opt-in) in **experiments.json**.

@elliot-nelson @D4N14L do you have opinions on naming/optimism?

File: libraries/rush-lib/src/api/LastInstallFlag.ts [hunk 2]
```diff
@@ -80,6 +82,29 @@ export class LastInstallFlag {
                 `New Path: ${newStorePath}`
```
Comment (octogonz) [inline libraries/rush-lib/src/api/LastInstallFlag.ts:82]: I noticed the **last-install.flag** data structure is collecting quite a lot of settings. Mine looks like this:

```js
{
  "node": "14.19.3",
  "packageManager": "pnpm",
  "packageManagerVersion": "6.29.1",
  "rushJsonFolder": "C:\\Git\\rushstack",
  "storePath": "C:\\Git\\rushstack\\common\\temp\\pnpm-store",
  "workspaces": true,
  "selectedProjectNames": [
    "@microsoft/rush-lib",
    "@rushstack/heft-config-file",
    "@rushstack/node-core-library",
    "@rushstack/package-deps-hash",
    "@rushstack/rig-package",
    "@rushstack/stream-collator",
    "@rushstack/terminal",
    "@rushstack/ts-command-line",
    "@rushstack/eslint-config",
    "@rushstack/heft",
    "@rushstack/heft-node-rig",
    "@rushstack/eslint-patch",
    "@rushstack/eslint-plugin",
    "@rushstack/eslint-plugin-packlets",
    "@rushstack/eslint-plugin-security",
    "@microsoft/api-extractor",
    "@rushstack/heft-jest-plugin",
    "@rushstack/tree-pattern",
    "@microsoft/api-extractor-model"
  ],
  "ignoreScripts": true
}
```

It would be helpful to define a proper TypeScript interface for this object instead of relying on `JsonObject` (i.e. `any`). (This is code cleanup, not directly part of your feature.)

File: libraries/rush-lib/src/api/LastInstallFlag.ts [hunk 3]
```diff
@@ -80,6 +82,29 @@ export class LastInstallFlag {
...
             );
```
Comment (octogonz) [inline libraries/rush-lib/src/api/LastInstallFlag.ts:83]: I was surprised to find that prior to this PR, Rush deletes **last-install.flag** flag when doing a subset install.

For example, if I run `rush install` twice in a row, the second time it prints `Installation is already up-to-date.` and skips invoking PNPM entirely.  Whereas (prior to this PR) if I run `rush install --to library` twice in a row, it would invoke PNPM both times, because **last-install.flag** flag is not created with `--to`.

This PR improves the behavior: If I run `rush install --to library` twice, the second time it does print `Installation is already up-to-date.` and skip `pnpm install` (although it runs `pnpm rebuild --pending` both times, by design).  That seems like an improvement. 👍

@D4N14L @iclanton Do you remember why this optimization wasn't implemented previously?  Is adding `"selectedProjectNames"` to **last-install.flag** flag sufficient to guarantee a valid install?  Are there any interesting edge cases we should test?

File: libraries/rush-lib/src/logic/base/BaseInstallManager.ts (hunk 1 above)
Comment (octogonz) [reply libraries/rush-lib/src/logic/base/BaseInstallManager.ts:193]: @chengcyber pointed out that `postinstall` doesn't have any special status among the sequence of `preinstall`, `install`, `postinstall`, `preprepare` ...  So maybe we can call them "installation scripts" (following the naming of `pnpm install --ignore-scripts`).

We discussed it a bit and liked the names  `deferredInstallationScripts` or `runInstallationScriptsSeparately`. I don't have a strong opinion.

File: libraries/rush-lib/src/api/LastInstallFlag.ts (hunk 3 above)
Comment (dmichon-msft) [reply libraries/rush-lib/src/api/LastInstallFlag.ts:105]: As long as the options to the previous install are an exact match, it should be fine to allow reuse. Adding or removing projects should cause a re-install, though.

Comment (dmichon-msft) [reply libraries/rush-lib/src/api/LastInstallFlag.ts:105]: The reason _removing_ projects should cause a reinstall is because, in the presence of hoisting (i.e. unless hoist=false in .npmrc), a filtered install and a full install will have different packages visible to the dependencies of the selected projects. If we want to enforce use of `--purge` to reproduce the behavior of a clean install, then we could potentially allow subset installs to be a no-op.

File: libraries/rush-lib/src/api/base/BaseFlag.ts [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+import * as path from 'path';
+import { FileSystem, Import, JsonFile, JsonObject } from '@rushstack/node-core-library';
+
+const lodash: typeof import('lodash') = Import.lazy('lodash', require);
+
+/**
+ * A base class for flag file.
+ * @internal
+ */
+export class BaseFlag {
```
Comment (dmichon-msft) [inline libraries/rush-lib/src/api/base/BaseFlag.ts:13]: Use a generic parameter for the type of `state`

File: libraries/rush-lib/src/api/LastInstallFlag.ts [hunk 5]
```diff
@@ -80,6 +104,29 @@ export class LastInstallFlag {
...
           }
+
+          // check ignoreScripts
+          if (newState.ignoreScripts !== oldState.ignoreScripts) {
+            return false;
+          } else {
+            // full install
+            if (newState.selectedProjectNames === oldState.selectedProjectNames) {
```
Comment (dmichon-msft) [inline libraries/rush-lib/src/api/LastInstallFlag.ts:113]: For clarity, `!newState.selectedProjectNames && !oldState.selectedProjectNames`; since if the value is an array they will always be inequal.

File: libraries/rush-lib/src/cli/actions/InstallAction.ts [hunk 6]
```diff
@@ -49,13 +52,33 @@ export class InstallAction extends BaseInstallAction {
...
     const terminal: Terminal = new Terminal(new ConsoleTerminalProvider());
+
+    const pnpmFilterArguments: string[] = await this._selectionParameters!.getPnpmFilterArgumentsAsync(
+      terminal
+    );
+
+    let selectedProjects: Set<RushConfigurationProject> | undefined;
+    if (pnpmFilterArguments.length > 0) {
+      selectedProjects = await this._selectionParameters!.getSelectedProjectsAsync(terminal);
+    }
```
Comment (dmichon-msft) [inline libraries/rush-lib/src/cli/actions/InstallAction.ts:76]: If you need both of these at once going forwards it makes sense to update the `getPnpmFilterArgumentsAsync` API to also return the selectedProjects to avoid duplicating work.

File: libraries/rush-lib/src/cli/actions/UpdateAction.ts [hunk 7]
```diff
@@ -54,6 +55,14 @@ export class UpdateAction extends BaseInstallAction {
         " to process the shrinkwrap file.  This will also update your shrinkwrap file with Rush's fixups." +
         ' (To minimize shrinkwrap churn, these fixups are normally performed only in the temporary folder.)'
     });
+    this._ignoreScriptsParameter = this.defineFlagParameter({
+      parameterLongName: '--ignore-scripts',
+      description:
+        'Do not execute any install lifecycle scripts specified in package.json files and its dependencies' +
```
Comment (D4N14L) [inline libraries/rush-lib/src/cli/actions/UpdateAction.ts:61]: Same here

File: libraries/rush-lib/src/logic/base/BaseInstallManager.ts [hunk 8]
```diff
@@ -154,11 +177,26 @@ export abstract class BaseInstallManager {
...
+
+    if (isFilteredInstall && !this.options.selectedProjects) {
+      // This should never even happen
+      throw new Error(`Missing selectedProjects when filtered install`);
```
Comment (D4N14L) [inline libraries/rush-lib/src/logic/base/BaseInstallManager.ts:197]: If it shouldn't ever happen, that's a bug. Throw an `InternalError`

File: libraries/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts [hunk 9]
```diff
@@ -27,6 +27,13 @@ import { BaseProjectShrinkwrapFile } from '../base/BaseProjectShrinkwrapFile';
  * This class implements common logic between "rush install" and "rush update".
  */
 export class WorkspaceInstallManager extends BaseInstallManager {
+  // Example: "C:\MyRepo\common\temp\npm-local\node_modules\.bin\npm"
+  private _packageManagerFilename: string = this.rushConfiguration.packageManagerToolFilename;
```
Comment (D4N14L) [inline libraries/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts:31]: I would prefer to set private variables that aren't static values/that reference `this` in the constructor (same goes for `_packageManagerEnv`), or just use the variable directly by referencing `this.rushConfiguration.packageManagerToolFilename`

File: libraries/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts [hunk 10]
```diff
@@ -414,6 +422,54 @@ export class WorkspaceInstallManager extends BaseInstallManager {
...
+      if (
+        this.rushConfiguration.pnpmOptions.pnpmStore === 'local' ||
+        EnvironmentConfiguration.pnpmStorePathOverride
+      ) {
+        // --store is not supported to pnpm rebuild... So, config.store is used here
+        rebuildArgs.push(`--config.storeDir=${this.rushConfiguration.pnpmOptions.pnpmStorePath}`);
+      }
+
+      if (this.options.collectLogFile) {
+        rebuildArgs.push('--reporter', 'ndjson');
```
Comment (D4N14L) [inline libraries/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts:439]: Will the logfile overwrite the existing log? Or is this to output logs into the console, which we subsequently append into the log we write?

File: libraries/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts [hunk 11]
```diff
@@ -414,6 +422,54 @@ export class WorkspaceInstallManager extends BaseInstallManager {
 
     // TODO: Remove when "rush link" and "rush unlink" are deprecated
     LastLinkFlagFactory.getCommonTempFlag(this.rushConfiguration).create();
+
+    // Stage 2 rebuild pending
+    if (this.twoStageInstall && !this.options.ignoreScripts) {
+      const rebuildArgs: string[] = ['rebuild', '--pending', '--recursive'];
+
+      if (
```
Comment (D4N14L) [inline libraries/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts:430]: It feels like a lot of this is duplicate from the method we use to append the default parameters. Maybe these can be made into a common method?

File: libraries/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts [hunk 12]
```diff
@@ -372,6 +376,10 @@ export class WorkspaceInstallManager extends BaseInstallManager {
   }
 
   protected async postInstallAsync(): Promise<void> {
+    if (this.twoStageInstall) {
+      this.commonTempInstallFlag.saveIfModified();
```
Comment (D4N14L) [inline libraries/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts:380]: The use of this flag file feels like it's getting muddled by this change. It used to mean "everything in the repo is installed and ready to use", now it means "some subset of the repo is installed and possibly the post-install scripts have been run".

It would make sense to use this file to do more specific state tracking (I like that you're including the scope in the file now). We should also care about running the post-install scripts though since not running them _could_ leave your repo in a less-usable state. Maybe we should check this file, and if `ignoreScripts` is `true` and `this.options.ignoreScripts` is false, we execute this stage, then write `ignoreScripts: false` in the flag file?

File: libraries/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts [hunk 13]
```diff
@@ -414,6 +422,54 @@ export class WorkspaceInstallManager extends BaseInstallManager {
...
+          args: rebuildArgs,
+          workingDirectory: this.rushConfiguration.commonTempFolder,
+          environment: this._packageManagerEnv,
+          suppressOutput: false
+        });
+      } catch (err) {
+        throw new Error(`Encounter an error when running install lifecycle scripts. error: ${err.message}`);
+      } finally {
+        // Always save after pnpm rebuild to update timestamp of last install flag file.
+        this.commonTempInstallFlag.create();
```
Comment (D4N14L) [inline libraries/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts:470]: Previously, this flag would only exist if there was a successful install. Use should be updated here to keep in line with what I describe above (update the flag for `ignoreScripts` in the case that the post-install scripts run + succeed, or leave as-is and use it to skip the install but use to execute post-install scripts.

File: libraries/rush-lib/src/logic/base/BaseInstallManager.ts (hunk 1 above)
Comment (chengcyber) [reply libraries/rush-lib/src/logic/base/BaseInstallManager.ts]: renaming "twoStageInstall" to "deferredInstallationScripts".

File: libraries/rush-lib/src/api/LastInstallFlag.ts (hunk 2 above)
Comment (chengcyber) [reply libraries/rush-lib/src/api/LastInstallFlag.ts:111]: A new `BaseFlag` class with generic type comes into rescue, and `LastInstallFlag` extends `BaseFlag<ILastInstallFlagJson>` now

File: libraries/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts (hunk 10 above)
Comment (chengcyber) [reply libraries/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts:439]: later one. :)

File: libraries/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts (hunk 12 above)
Comment (chengcyber) [reply libraries/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts:380]: Actually, there are only two cases: 1. `ignoreScripts: true` (deferredInstallationScripts or '--ignore-scripts' specified in CLI parameter); 2. In other cases, `ignoreScript` is missing in the flag file.

Comment (chengcyber) [reply libraries/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts:380]: It kind of complicated. Say we have run `rush install --ignore-scripts --to rush`, the flag file records:

```
"selectedProjects": ["<other_deps>", ..., "rush-lib", "rush"]
"ignoreScripts": true
```

Now, if i run `rush install --to rush-lib`, the stage 1 `pnpm install --ignore-scripts` should be skipped, since `--to rush-lib` is a subset of `--to rush`. And Rush just run stage 2 `pnpm rebuild --filter rush-lib...`. Now we can **not** mark `ignoreScripts: false` in the flag file. Since the `selectedProjects` contains "rush", which is out of range of `--filter rush-lib...`

One possible solution maybe is to record more information in one selected project in the flag file, use the above case as an example, the flag file will be

```
"selectedProjects": [
// ... other deps
{
  packageName: 'rush-lib',
  ignoreScripts: false,
},
{
  packageName: 'rush',
  ignoreScripts: true,  // <--- because we run "pnpm rebuild --pending --filter rush-lib..." which excludes "rush"
}
]
```

In this way, in `postinstallAsync` we need calculate `ignoreScripts` based on all selectedProjects to figure out whether we can skip "pnpm rebuild --pending"...

Do you think it's worth the efforts to implement this logic? For now, i choose an easy implementation which run "pnpm rebuild --pending [--filter <project...>]" everytime. `--pending` let pnpm helps decide whether there is any script we need to run.

File: libraries/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts (hunk 13 above)
Comment (chengcyber) [reply libraries/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts:464]: The reason why invoking creating last install flag here: `canSkipInstall` compares a timestamp(mtime of this flag file) with mtime of node_modules folder. And no matter `pnpm rebuild --pending` runs successfully or not, `node_modules` folder is touched, i.e. mtime updated. So we need to update the mtime of this flag file to make the `canSkipInstall` works again.

File: libraries/rush-lib/src/api/LastInstallFlag.ts [hunk 14]
```diff
@@ -80,6 +104,29 @@ export class LastInstallFlag {
...
+              return true;
```
[the remark this answers is not in the document]
Comment (chengcyber) [reply libraries/rush-lib/src/api/LastInstallFlag.ts:121]: I agree with you. but i have no idea how to do it here. Could you give some suggestions on how to change this part of code for better maintenance.

File: libraries/rush-lib/assets/rush-init/common/config/rush/experiments.json [hunk 15]
```diff
@@ -40,5 +40,11 @@
...
    */
-  /*[LINE "HYPOTHETICAL"]*/ "phasedCommands": true
-}
+  /*[LINE "HYPOTHETICAL"]*/ "phasedCommands": true,
+
+  /**
+   * If true, rush install or rush update implicitly specify --ignore-scripts during pnpm install,
+   * and run install lifecycle scripts by pnpm rebuild --pending after pnpm install successfully.
+   */
+  /*[LINE "HYPOTHETICAL"]*/ "deferredInstallationScripts": true
```
Comment (iclanton) [inline libraries/rush-lib/assets/rush-init/common/config/rush/experiments.json:49]: Will this eventually be our recommended approach?

File: libraries/rush-lib/src/api/LastInstallFlag.ts [hunk 16]
```diff
@@ -1,39 +1,70 @@
...
 
+/**
+ * This represents the JSON data structure for the "last-install.flag" file.
+ * @internal
+ */
+export interface ILastInstallFlagJson {
+  /**
+   * Current node version
+   */
+  node: string;
```
Comment (iclanton) [inline libraries/rush-lib/src/api/LastInstallFlag.ts:22]:
```suggestion
  nodeVersion: string;
```

File: libraries/rush-lib/src/api/LastInstallFlag.ts [hunk 17]
```diff
@@ -1,39 +1,70 @@
...
+   */
+  packageManagerVersion: string;
+  /**
+   * Current rush json folder
+   */
+  rushJsonFolder: string;
+  /**
+   * The content of package.json, used in the flag file of autoinstaller
+   */
+  packageJson?: IPackageJson;
```
Comment (iclanton) [inline libraries/rush-lib/src/api/LastInstallFlag.ts:38]:
```suggestion
  autoinstallerPackageJson?: IPackageJson;
```

File: libraries/rush-lib/src/api/LastInstallFlag.ts [hunk 18]
```diff
@@ -1,39 +1,70 @@
...
+  /**
+   * Same with pnpmOptions.pnpmStorePath in rush.json
+   */
+  storePath?: string;
```
Comment (iclanton) [inline libraries/rush-lib/src/api/LastInstallFlag.ts:42]:
```suggestion
  pnpmStorePath?: string;
```

File: libraries/rush-lib/src/api/LastInstallFlag.ts [hunk 19]
```diff
@@ -1,39 +1,70 @@
...
+  /**
+   * True when "useWorkspaces" is true in rush.json
+   */
+  workspaces?: true;
```
Comment (iclanton) [inline libraries/rush-lib/src/api/LastInstallFlag.ts:46]:
```suggestion
  useWorkspaces?: true;
```

File: libraries/rush-lib/src/api/LastInstallFlag.ts [hunk 20]
```diff
@@ -1,39 +1,70 @@
...
+  /**
+   * True when user explicitly specify "--ignore-scripts" CLI parameter or deferredInstallationScripts
+   */
+  ignoreScripts?: true;
+  /**
+   * When specified, it is a list of selected projects during partial install
+   * It is undefined when full install
+   */
+  selectedProjectNames?: string[];
```
Comment (iclanton) [inline libraries/rush-lib/src/api/LastInstallFlag.ts:55]:
```suggestion
  partialInstallSelectedProjectNames?: string[];
```

File: libraries/rush-lib/src/cli/actions/InstallAction.ts [hunk 21]
```diff
@@ -49,13 +50,29 @@ export class InstallAction extends BaseInstallAction {
...
+    this._ignoreScriptsParameter = this.defineFlagParameter({
+      parameterLongName: '--ignore-scripts',
+      description:
+        'Do not execute any install lifecycle scripts specified in package.json files and its' +
+        ' dependencies when "rush install". Running with this flag leaves your installation in a uncompleted' +
+        ' state, you need to run without this flag again to complete a full installation. Meanwhile, it makes' +
+        ' your installing faster. Later, you can run "rush install" to run all ignored scripts. Moreover, you' +
+        ' can partial install such as "rush install --to <package>" to run ignored scripts of the dependencies' +
+        ' of the selected projects.'
+    });
```
Comment (iclanton) [inline libraries/rush-lib/src/cli/actions/InstallAction.ts:63]: Is this only supported when using PNPM? We should either explicitly mention that and error if someone uses this flag in a non-pnpm repo, or just not include this flag at all in non-pnpm repos.

File: libraries/rush-lib/src/cli/actions/UpdateAction.ts [hunk 22]
```diff
@@ -54,6 +55,16 @@ export class UpdateAction extends BaseInstallAction {
         " to process the shrinkwrap file.  This will also update your shrinkwrap file with Rush's fixups." +
         ' (To minimize shrinkwrap churn, these fixups are normally performed only in the temporary folder.)'
     });
+    this._ignoreScriptsParameter = this.defineFlagParameter({
```
Comment (iclanton) [inline libraries/rush-lib/src/cli/actions/UpdateAction.ts:58]: Same here as the install version of the flag.

File: libraries/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts [hunk 23]
```diff
@@ -414,6 +418,56 @@ export class WorkspaceInstallManager extends BaseInstallManager {
 
     // TODO: Remove when "rush link" and "rush unlink" are deprecated
     LastLinkFlagFactory.getCommonTempFlag(this.rushConfiguration).create();
+
+    // Stage 2 rebuild pending
```
Comment (iclanton) [inline libraries/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts:422]: Can you include a `// Stage 1` comment where stage 1 happens?

File: libraries/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts [hunk 24]
```diff
@@ -433,4 +487,27 @@ export class WorkspaceInstallManager extends BaseInstallManager {
       }
     }
   }
+
+  /**
+   * Used when invoking pnpm rebuild for running deferred installation scripts.
+   */
+  protected pushRebuildArgs(args: string[]): void {
```
Comment (iclanton) [inline libraries/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts:494]: "Rebuild" has a specific meaning in the current version of Rush. Can we use a different word here?

Comment (octogonz) [reply libraries/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts:494]: We could call it `pushPnpmRebuildArgs()`

File: libraries/rush-lib/assets/rush-init/common/config/rush/experiments.json (hunk 15 above)
Comment (octogonz) [reply libraries/rush-lib/assets/rush-init/common/config/rush/experiments.json:49]: Yes, I think so

File: libraries/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts (hunk 24 above)
Comment (iclanton) [reply libraries/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts:494]: That kinda makes it sound like we're rebuilding PNPM.

Can we just call this operation "postInstallScripts" everywhere? So this could be `pushPnpmPostInstallScriptsArgs`

Comment (octogonz) [reply libraries/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts:494]: We're now calling them "installation scripts" because `postinstall` is just one of several lifecycle events that happen during installation (`preinstall`, `install`, `prepare`, etc).

That said, @iclanton these are CLI arguments for the actual shell command `pnpm rebuild`.  PNPM's naming could be more intuitive, but `rebuild` is the name of the command.  It doesn't seem helpful for our source code to refer to it using a different name.

We could qualify it a bit as `pushPnpmRebuildCommandArgs`... and maybe add a code comment:

```ts
 /**
   * Used when invoking the "pnpm rebuild" command that runs installation scripts of NPM dependencies.
   * NOTE: Despite the name, "pnpm rebuild" is purely an installation action -- it does not build Rush projects
   * in the manner of "rush rebuild" or "rush build".
   */
  protected pushPnpmRebuildCommandArgs(args: string[]): void {
```

File: libraries/rush-lib/src/cli/actions/InstallAction.ts (hunk 21 above)
Comment (octogonz) [reply libraries/rush-lib/src/cli/actions/InstallAction.ts:63]: > just not include this flag at all in non-pnpm repos.

Better to include the flag and throw an error saying `the "--ignore-scripts" parameter is not supported with your selected package manager`.

It is an infuriating user experience for CLI parameters to mysteriously disappear based on context. (*"I SWEAR Rush has this feature! Did I forget the name of it? Am I losing my mind?!"*  *"No seriously, look, the docs even mention it!"*  *"Did they remove the feature? Why would they do that?!"*) 😆

File: libraries/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts (hunk 12 above)
Comment (D4N14L) [reply libraries/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts:376]: Good question. I think keeping more detailed info on the selected projects does make sense... it allows you to be very specific about the work you're performing and when you want to perform that work. Another option IMO would be to execute the install from scratch again (ie. running `pnpm install --ignore-scripts ...` then `pnpm rebuild ...`) whenever non-scoping arguments to the install have changed.

Basically, I feel like we either need to commit to tracking each project install on a per-project basis, or we need to treat all other options as something that could affect all projects and start from scratch.

File: libraries/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts (hunk 13 above)
Comment (D4N14L) [reply libraries/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts:468]: Right, but skipping an install in the event of an error doesn't make sense. It should be updated only after the command is successfully executed. If the command fails, then the state of the repo is unknown and the last-install flag should probably be deleted.

Comment (chengcyber) [reply libraries/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts:468]: It's kindof fight against the main point of the proposal, faster retries on postinstall scripts.

Imagine

```
pnpm install --ignore-scripts --filter <a...>
```

works

```
pnpm rebuild --pending --filter <a...>
```

If `pnpm rebuild` fails, Rush.js doesn't update the last-install.flag. The "canSkipInstall" will never return true in this case. which triggers a "pnpm install --ignore-scripts --filter <a...>" again...

but, the main point is to make "canSkipInstall" returns true in this case, and only the stage 2 command("pnpm rebuild" ) runs, which is the "faster retires" part...

File: libraries/rush-lib/src/api/LastInstallFlag.ts (hunk 16 above)
Comment (chengcyber) [reply libraries/rush-lib/src/api/LastInstallFlag.ts:22]: Hi, @iclanton . Do you need to rename these fields in this PR? It seems that most fields like "node", "packageJson", "storePath", "workspaces" are existing fields in current last install flag.

File: libraries/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts (hunk 13 above)
Comment (D4N14L) [reply libraries/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts:468]: Then in that case it should write the flag file twice, once after performing the install (with `ignoreScripts: true`), then once after running the post-install scripts (with `ignoreScripts: false`). We shouldn't be saying "yes, everything is installed correctly" regardless of success or failure, that defeats the entire purpose of the flag file.

Comment (octogonz) [reply libraries/rush-lib/src/logic/installManager/WorkspaceInstallManager.ts:468]: > in that case it should write the flag file twice, once after performing the install

+1  Rush install is meant to be transactionally safe.  It is a common optimization for CI agents to reuse the same `git clone` working folder from a previous job. And it is very common for the previous `rush install` to have been terminated because someone pushed an update to their PR branch, which causes the previous CI job to get immediately cancelled and restarted.

**Comment by chengcyber** [comment]
Updates 2022/8/7:

1. rebase latest main branch
2. rename pushRebuildArgs to pushPnpmRebuildCommandArgs
3. add a comment on Stage 1
4. throw error when specifying --ignore-scripts in non-pnpm repos
5. refactor last-install flag and add a individual `isSelectedProjectInstalled` method

**Comment by chengcyber** [comment]
Update: Sync this branch with the latest main branch

File: common/changes/@microsoft/rush/feat-mitigation-postinstall-failures_2022-07-14-07-36.json [hunk 25]
```diff
@@ -0,0 +1,10 @@
+{
+  "changes": [
+    {
+      "packageName": "@microsoft/rush",
+      "comment": "Supports deferredInstallationScripts behind a experimental flag, which makes faster retires for install lifecycle scripts",
```
[the remark this answers is not in the document]
Comment (chengcyber) [reply common/changes/@microsoft/rush/feat-mitigation-postinstall-failures_2022-07-14-07-36.json:5]: Actually, there is no `--defereredInstallScripts` CLI flag. When turn on the experimental flag, post install scripts are invoked after `pnpm install` during "rush install/update"...

**Comment by chengcyber** [comment]
Update:

1. merge the latest main branch
2. update change log comment
